### DISCUSSION TITLE
How to make nav links close the collapse menu?

### DISCUSSION THREAD

**Discussion by DeanParkin** [twbs/bootstrap#36786] [Q&A] [answered]
I'm using Bootstrap 5 and Next.js and I'm trying to collapse the Navbar menu when I click a link in mobile view. I've tried a few different methods but I'm hitting a few roadblocks and I can't find any documentation on preferred methods. 

I've tried using the classlist.remove('show') and this makes the toggle method very jumpy and effects the toggle process.

I've tried using ( data-bs-toggle"collapse" data-bs-target="#navbarToggler" ) as properties on the links and this does collapse the menu but the links then blink in and out when the collapse menu is past its breakpoint and is not in the collapse menu. 

I'm trying now to access the toggle method for that collapse by using the onClick property on the link, and I can't seem to get that to work at all. my method is currently:

onClick={() => {
                      const tog = document.getElementById("navbarToggler");
                      if (
                        window.innerWidth < 768 &&
                        tog.classList.contains("show")
                      ) {
                        tog.collapse({ toggle: false }).collapse("hide");
                      } else {
                        console.log("no");
                      }
                    }}

Not sure what to do but wondering if anyone here has some experience with this and a preferred go-to method, Thanks for taking the time to read this

**Comment by GeoSot** [comment]
Probably you are searching something like

```js
const tog = document.getElementById("navbarToggler")
if ( window.innerWidth < 768 &&  tog.classList.contains("show") ) {    
    const instance=  bootstrap.collapse.getOrCreateInstance(tog)
    instance.hide()
} else {
     console.log("no");
}
```

**Reply by DeanParkin** [reply]
because I'm using Next.js it's telling me:

Unhandled Runtime Error
ReferenceError: bootstrap is not defined

**Reply by GeoSot** [reply]
I suppose you have to find how you have exported BS javascript (namespaced or not)
some examples here :wink: 
https://www.webdevtutor.net/blog/how-to-add-bootstrap-5-to-next-js-app
https://blog.logrocket.com/handling-bootstrap-integration-next-js/

**Comment by DeanParkin** [comment, marked as answer]
I've managed to do it, it's probably not the soundest way to do it in react but it's working I ended up doing something along the lines of this:

useEffect(() => {
    let size = window.innerWidth;
    let navLinks = document.querySelectorAll(".nav-link");

    const updateWindowDimensions = () => {
      size = window.innerWidth;
    };
    window.addEventListener("resize", () => {
      updateWindowDimensions();
      if (size > 992) {
        navLinks.forEach((link) => {
          link.removeAttribute("data-bs-toggle");
          link.removeAttribute("data-bs-target");
        });
        console.log("removed attributes");
      } else {
        navLinks.forEach((link) => {
          link.setAttribute("data-bs-toggle", "collapse");
          link.setAttribute("data-bs-target", "#navbarToggler");
        });
        console.log("set attributes");
      }
    });
  }, []);

  const handleClick = (e) => {
    e.preventDefault();
    router.push(e.target.href);
  };

**Reply by DeanParkin** [reply]
Cant F*$*%(^ believe it, been trying to do this for weeks, spent an entire day on it yesterday, finally pipe up and ask a question today and within an hour I've managed to crack it, Sods Law, Thank you for your help Geo

**Reply by arturhonores** [reply]
Dear @DeanParkin 
I have the same problem, but I don't know how I can add your solution to my script, could you provide me with the code, I can't find a solution in other forums.

I only found this, but it doesn't work for me:
`const navLinks = document.querySelectorAll(".nav-item");
const menuToggle = document.getElementById("navbarSupportedContent");
const bsCollapse = new bootstrap.Collapse(menuToggle);
navLinks.forEach((l) => {
  l.addEventListener("click", () => {
    bsCollapse.toggle();
  });
});`

**Reply by DeanParkin** [reply]
try this:

```
//collect all your nav links
let navLinks = document.querySelectorAll(".nav-link");

// variable to store the size of the window
let size = window.innerWidth;

// function to toggle the data-bs attributes depending on the breakpoint set
function setAttr() {

  // 992 is the md breakpoint
  if (size > 992) {
    navLinks.forEach((link) => {
      link.removeAttribute("data-bs-toggle");
      link.removeAttribute("data-bs-target");
    });
  } else {
    navLinks.forEach((link) => {
      link.setAttribute("data-bs-toggle", "collapse");
      // #navbarToggler is the id of the navbar collapse
      link.setAttribute("data-bs-target", "#navbarToggler");
    });
  }
}

//function to update the size variable 
const updateWindowDimensions = () => {
  size = window.innerWidth;
};

// event listener to call the function to update the size variable and check the breakpoint
window.addEventListener("resize", () => {
  updateWindowDimensions();
  setAttr();
});

// call the function to set the attributes
setAttr();

// call on the document on load just for good measure
document.onload = () => {
  setAttr();
};
```

**Reply by arturhonores** [reply]
@DeanParkin 

This code has worked for me

```
const navLinks = document.querySelectorAll('.nav-item')
const menuToggle = document.getElementById('navbarSupportedContent')
const bsCollapse = new bootstrap.Collapse(menuToggle, {toggle:false})
navLinks.forEach((l) => {l.addEventListener('click', () => { bsCollapse.toggle() })
})
```

**Reply by DeanParkin** [reply]
I tried using that on next js, and I had no luck with it, that's why I used that code I sent you, I applied it in a useEffect statement, not sure if there is another way to do it but that worked for me.

**Comment by tungphan995** [comment]
You are the best.

**Reply by arturhonores** [reply]
@tungphan995 
this code works perfectly

```
const navLinks = document.querySelectorAll('.nav-item')
const menuToggle = document.getElementById('navbarSupportedContent')
const bsCollapse = new bootstrap.Collapse(menuToggle, {toggle:false})
navLinks.forEach((l) => {l.addEventListener('click', () => { bsCollapse.toggle() })
})
```

**Reply by aselemidivine** [reply]
This works perfectly fine. Thanks a lot.

**Reply by Ayushpatil3103** [reply]
hey this code doesn't run second time?we have to refresh

**Reply by Pholarin** [reply]
It worked for me too. Thanks

**Comment by theboyofdream** [comment]
```
const menuToggle = document.getElementById('navbarCollapse')
    const navLinks = menuToggle.querySelectorAll('.nav-item')
    const bsCollapse = new bootstrap.Collapse(menuToggle, {toggle:false})
    navLinks.forEach((l) => {
      l.addEventListener('click', () => {
        if(window.innerWidth < 992) { bsCollapse.toggle() }
      })
    })
```
Thanks for help.
